packages/swing-store/src/snapStore.js

```diff
@@ -360,20 +365,93 @@ export function makeSnapStore(
   `);
   sqlGetSnapshotList.pluck(true);
 
+  const sqlGetSnapshotListLimited = db.prepare(`
+    SELECT snapPos
+    FROM snapshots
+    WHERE vatID = ?
+    ORDER BY snapPos ASC
+    LIMIT ?
+  `);
+  sqlGetSnapshotListLimited.pluck(true);
+
+  /**
+   * @param {string} vatID
+   * @returns {boolean}
+   */
+  function hasSnapshots(vatID) {
+    return !!sqlGetSnapshotListLimited.all(vatID, 1).length;
+  }
+
   /**
-   * Delete all snapshots for a given vat (for use when, e.g., a vat is terminated)
    *
    * @param {string} vatID
+   * @param {number} budget
+   * @returns {{ done: boolean, cleanups: number }}
```

Comment:
It does, but I didn't find a way to take advantage of that fact.

An actual `function*` generator wouldn't work, of course, because the process can be killed and a new process started while the deletion is going on, and a real generator would lose state when the application is rebooted.

And, I think changing the function signature to match that of a normal `function*` generator is only an improvement if the caller gets to use for..of syntax, but as long as the snapStore function is doing internal iteration (deleting more than one thing per call), the vatKeeper.js `deleteSnapshotsAndTranscripts()` caller is only going to call it once per block (per terminated vat), so there's no good place for a `for..of` loop. (the real loop is higher up, with one iteration per block).

To get one, we'd need to change snapStore's `deleteSomeVatSnapshots` into `maybeDeleteOneVatSnapshot`, to delete at most one per call, and then have vatKeeper's `deleteSnapshotsAndTranscripts()` use a `for..of` loop. We'd still need to return whether a cleanup was done or not, and have the caller accumulate them, so `deleteSnapshotsAndTranscripts` knows when to switch from snapshots to transcripts. `maybeDeleteOneVatSnapshot` would always make one DB query (with a LIMIT 1) to get which snapshot to delete, if any. Then it either returns, or does a second DB query to delete the one row, and a third to `noteExport` the deletion, making the cost 3 small DB queries until all the snapshots are gone, then 1 small DB query each block until all the transcripts are gone (since we always check for remaining snapshots on the way to checking for transcript spans).

That's compared to the current cost (with a budget of 5) of one moderate-sized query every time (using LIMIT 5, returning anywhere from 0 to 5 rows), followed by 0 to 5 `noteExports`, maybe followed by a single `DELETE` query removing 1 to 5 rows at once.

And we'd need snapStore to expose `maybeDeleteOneVatSnapshot` separately from `deleteVatSnapshots` (unlimited), so the latter could to queries without `LIMIT` constraints, and delete everything in one shot.

In general, it de-amortizes the DB queries, because to make use of the iterator, we have to move responsibility for doing more than one deletion (per block) up into vatKeeper, which then can't give a hint to swingstore about how many deletions are coming up, so it could query them in a batch.